File: backtracking/090_subsets_ii-solved/solution.ts
/*
90. Subsets II
https://leetcode.com/problems/subsets-ii/

Sort first, then backtrack. Skip duplicates at the same recursion depth
(i > start && nums[i] === nums[i-1]) to avoid generating duplicate subsets.
*/

function subsetsWithDup(nums: number[]): number[][] {
    nums.sort((a, b) => a - b)
    const result: number[][] = []
    function backtrack(temp: number[], start: number) {
        result.push([...temp])
        for (let i = start; i < nums.length; i++) {
            if (i !== start && nums[i] === nums[i - 1]) continue;
            temp.push(nums[i])
            backtrack(temp, i + 1)
            temp.pop()
        }
    }

    backtrack([], 0)

    return result
}

File: backtracking/090_subsets_ii-solved/solution.py
"""
90. Subsets II
https://leetcode.com/problems/subsets-ii/

Sort first, then backtrack. Skip duplicates at the same recursion depth
(i > start and nums[i] == nums[i-1]) to avoid generating duplicate subsets.
"""


def subsetsWithDup(nums: list[int]) -> list[list[int]]:
    nums.sort()
    result = []

    def backtrack(start, current):
        result.append(current[:])
        for i in range(start, len(nums)):
            if i > start and nums[i] == nums[i - 1]:
                continue
            current.append(nums[i])
            backtrack(i + 1, current)
            current.pop()

    backtrack(0, [])
    return result
